domain/model/service/service_test.go

```diff
@@ -1112,3 +1122,74 @@ func (s *serviceSuite) TestDefaultModelCloudNameAndCredential(c *gc.C) {
 		Name:  "test-cred",
 	})
 }
+
+// TestWatchActivatedModels verifies that WatchActivatedModels correctly sets up a watcher
+// that emits events for activated models when the watcher receives change events.
+func (s *serviceSuite) TestWatchActivatedModels(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+	ctx := context.Background()
+	svc := NewWatchableService(
+		s.mockState,
+		s.mockModelDeleter,
+		loggertesting.WrapCheckLog(c),
+		s.mockWatcherFactory,
+	)
+
+	activatedModelUUID1 := modeltesting.GenModelUUID(c)
+	activatedModelUUID2 := modeltesting.GenModelUUID(c)
+	activatedModelUUIDs := []coremodel.UUID{activatedModelUUID1, activatedModelUUID2}
+	activatedModelUUIDStrs := transform.Slice(activatedModelUUIDs, func(uuid coremodel.UUID) string {
+		return uuid.String()
+	})
+
+	// Set up necessary state function return values.
+	s.mockState.EXPECT().GetAllActivatedModelsUUIDQuery().Return(
+		"SELECT uuid from model WHERE activated = true",
+	)
+	s.mockState.EXPECT().GetActivatedModelUUIDs(gomock.Any(), gomock.Any()).Return(
+		activatedModelUUIDs, nil,
+	)
+
+	// Create change events for watcher mapper.
+	chEvent1 := changestreammock.NewMockChangeEvent(s.mockWatcherFactory.ctrl)
+	chEvent1.EXPECT().Changed().Return(
+		activatedModelUUID1.String(),
+	)
+
+	chEvent2 := changestreammock.NewMockChangeEvent(s.mockWatcherFactory.ctrl)
+	chEvent2.EXPECT().Changed().Return(
+		activatedModelUUID2.String(),
+	)
+
+	chEvent3 := changestreammock.NewMockChangeEvent(s.mockWatcherFactory.ctrl)
+	chEvent3.EXPECT().Changed().Return(
+		"unactivated model UUID",
+	)
+
+	// Change events received by the watcher mapper.
+	receivedChangeEvents := []changestream.ChangeEvent{chEvent1, chEvent2, chEvent3}
+	// Change events containing model UUIDs of activated models retrieved from the database.
+	expectedChangeEvents := []changestream.ChangeEvent{chEvent1, chEvent2}
+
+	changes := make(chan []string, 1)
+	changes <- activatedModelUUIDStrs
+
+	w := watchertest.NewMockStringsWatcher(changes)
+
+	mapper := getWatchActivatedModelsMapper(s.mockState)
+	s.mockWatcherFactory.EXPECT().NewNamespaceMapperWatcher("model", changestream.Changed,
+		gomock.Any(), gomock.Any()).DoAndReturn(
+		func(string, changestream.ChangeType, eventsource.NamespaceQuery, eventsource.Mapper) (watcher.Watcher[[]string], error) {
+			// Use service mapper to retrieve change events containing only model UUIDs of activated models.
+			changes, err := mapper(ctx, s.ControllerTxnRunner(), receivedChangeEvents)
```

Comment:
Don't know why you put these checks here.
The mapper is called inside `svc.WatchActivatedModels(ctx)` and the 
last `c.Check(<-watcher.Changes(), gc.DeepEquals, activatedModelUUIDStrs)` checks this already for you.

Reply:
Unless you really want to test the mapper, then you can create a new test for testing the mapper func. This is the one for testing the service method.

Reply:
Yup I wanted to test the mapper as I felt that it's the core functionality of the method. Yup I agree that putting them in the same test is confusing since the test is meant for the service method itself, I will refactor the mapper test to another test, thanks Kelvin :)